subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/DefaultFileSystemSnapshotter.java

```diff
@@ -208,20 +224,190 @@ public void appendToHasher(BuildCacheHasher hasher) {
+            buffer[bufferSize++] = deferred;
+            fileTreeElements.add(deferred);
+            if (bufferSize == BATCH_SIZE) {
+                flush();
+            }
+        }
+
+        private void visitFileDirect(final FileVisitDetails fileDetails) {
+            fileTreeElements.add(new RegularFileSnapshot(internPath(fileDetails.getFile()), fileDetails.getRelativePath(), false, fileSnapshot(fileDetails)));
+        }
+
+        public List<FileSnapshot> getElements() {
+            if (completed) {
+                return fileTreeElements;
+            }
+            flush();
+            int i = 0;
+            for (FileSnapshot element : fileTreeElements) {
+                if (element instanceof DeferredFileSnapshot) {
+                    fileTreeElements.set(i, ((DeferredFileSnapshot) element).getResult());
+                }
+                i++;
+            }
+            completed = true;
+            return fileTreeElements;
+
+        }
+
+        private void flush() {
+            if (buffer == null || bufferSize == 0) {
+                return;
+            }
+            if (bufferSize == 1) {
+                synchronousSnapshot();
+                return;
+            }
+            submitForConcurrentExecution();
+        }
+
+        private void synchronousSnapshot() {
+            buffer[0].run();
+            buffer[0] = null;
+            bufferSize = 0;
+        }
+
+        private void submitForConcurrentExecution() {
+            final Runnable[] tasks = buffer.clone();
+            final int len = bufferSize;
+            for (int i = 0; i < bufferSize; i++) {
+                buffer[i] = null;
+            }
+            bufferSize = 0;
+            executorService.submit(new Runnable() {
+                @Override
+                public void run() {
+                    List<Throwable> errors = null;
+                    for (int i = 0; i < len; i++) {
+                        try {
+                            tasks[i].run();
+                        } catch (Throwable err) {
+                            if (errors == null) {
+                                errors = new ArrayList<Throwable>();
+                            }
+                            errors.add(err);
+                        }
+                    }
+                    if (errors != null) {
+                        rethrow(errors);
+                    }
+                }
+            });
+        }
+    }
+
+    private static void rethrow(List<Throwable> errors) {
+        if (errors.size() == 1) {
+            throw UncheckedException.throwAsUncheckedException(errors.get(0));
+        } else {
+            throw new MultipleBuildFailures(errors);
+        }
+    }
+
+    @Override
+    public void close() throws IOException {
+        executorService.shutdown();
+    }
+
+    private class DeferredFileSnapshot implements FileSnapshot, Runnable {
```

Comment:
Like said above, I don't like that `DeferredFileSnapshot` implements `FileSnapshot`. Conceptually, it looks much more like a `Future<FileSnapshot>` (and a `Runnable`, as it already is). WDYT?